core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1984,6 +1984,42 @@ object Stream {
       self.reduce(S.combine(_, _))
 
     /**
+      * Repartitions the input with the function `p`. On each step `p` is applied
+      * to the input and all elements but the last of the resulting sequence
+      * are emitted. The last element is then appended to the next input using the
+      * Semigroup `S`.
+      *
+      * @example {{{
+      * scala> import cats.implicits._
+      * scala> Stream("Hel", "l", "o Wor", "ld").repartition(s => Segment.array(s.split(" "))).toList
+      * res0: List[String] = List(Hello, World)
+      * }}}
+      */
+    def repartition(p: O => Segment[O, Unit])(implicit S: Semigroup[O]): Stream[F,O] = {
+      def go(carry: Option[O], s: Stream[F,O]): Pull[F,O,Unit] = {
+        s.pull.uncons.flatMap {
+          case Some((ohd, otl)) =>
+            ohd.uncons1 match {
+              case Right((hhd,htl)) =>
+                val tl = Stream.segment(htl) ++ otl
+                val next = carry.fold(hhd)(c => S.combine(c, hhd))
+                val parts = p(next).toVector
+                parts.size match {
+                  case 0 => go(None, tl)
+                  case 1 => go(Some(parts.head), tl)
+                  case _ => Pull.output(Chunk.indexedSeq(parts.init)) >> go(Some(parts.last), tl)
+                }
+              case Left(_) =>
+                carry.fold[Pull[F,O,Unit]](Pull.done)(c => Pull.output1(c))
```

Comment:
If we reimplement this to avoid the `uncons1`, then you can ignore the following comment. However, I don't think we want to end the pull here -- we'd hit this case if an empty segment is emitted. I think the right thing to do here is to recurse, carrying over the carry in to the recursive call.

Reply:
You're right, it should skip the segment in this case.